lib/src/main/kotlin/app/cash/kfsm/Transitioner.kt

```diff
@@ -0,0 +1,47 @@
+package app.cash.kfsm
+
+import app.cash.quiver.extensions.ErrorOr
+import app.cash.quiver.extensions.flatTap
+import arrow.core.Either
+import arrow.core.flatMap
+import arrow.core.flatten
+import arrow.core.left
+import arrow.core.raise.either
+import arrow.core.right
+
+open class Transitioner<S : State>(
+  private val persist: (Value<S>) -> ErrorOr<Value<S>> = { it.right() }
+) {
+
+  open fun preHook(value: Value<S>): ErrorOr<Unit> = Unit.right()
+
+  open fun postHook(value: Value<S>): ErrorOr<Unit> = Unit.right()
```

Comment:
I was considering commenting here that some information about the previous state might be useful here, but on reflection that might encourage bad modelling.